packages/dds/tree/src/feature-libraries/schemaIndex.ts

```diff
@@ -15,13 +15,168 @@ import { SummaryTreeBuilder } from "@fluidframework/runtime-utils";
 import { Index, SummaryElement, SummaryElementParser, SummaryElementStringifier } from "../shared-tree-core";
 import { cachedValue, ICachedValue, recordDependency } from "../dependency-tracking";
 import { Delta } from "../tree";
-import { StoredSchemaRepository } from "../schema-stored";
+import {
+    FieldKindIdentifier, FieldSchema, GlobalFieldKey, LocalFieldKey, Named,
+    SchemaData, StoredSchemaRepository, TreeSchema, TreeSchemaIdentifier, ValueSchema,
+} from "../schema-stored";
+import { brand } from "../util";
 
 /**
  * The storage key for the blob in the summary containing schema data
  */
 const schemaBlobKey = "SchemaBlob";
 
+const schemaStringKey = "SchemaString";
+
+const version = "1.0.0" as const;
+
+/**
+ * Format for encoding as json.
+ *
+ * For consistency all lists are sorted and undefined values are omitted.
+ *
+ * This chooses to use lists of named objects instead of maps:
+ * this choice is somewhat arbitrary, but avoids user data being used as object keys,
+ * which can sometimes be an issue (for example handling that for "__proto__" can require care).
+ */
+interface Format {
+    version: typeof version;
+    treeSchema: TreeSchemaFormat[];
+    globalFieldSchema: NamedFieldSchemaFormat[];
+}
+
+interface TreeSchemaFormat {
+    name: TreeSchemaIdentifier;
+    localFields: NamedFieldSchemaFormat[];
+    globalFields: GlobalFieldKey[];
+    extraLocalFields: FieldSchemaFormat;
+    extraGlobalFields: boolean;
+    value: ValueSchema;
+}
+
+type NamedFieldSchemaFormat = FieldSchemaFormat & Named<string>;
+
+interface FieldSchemaFormat {
+    kind: FieldKindIdentifier;
+    types?: TreeSchemaIdentifier[];
+}
+
+function encodeRepo(repo: SchemaData): Format {
+    const treeSchema: TreeSchemaFormat[] = [];
+    const globalFieldSchema: NamedFieldSchemaFormat[] = [];
+    for (const [name, schema] of repo.treeSchema) {
+        treeSchema.push(encodeTree(name, schema));
+    }
+    for (const [name, schema] of repo.globalFieldSchema) {
+        globalFieldSchema.push(encodeNamedField(name, schema));
+    }
+    treeSchema.sort(compareNamed);
+    globalFieldSchema.sort(compareNamed);
+    return {
+        version,
+        treeSchema,
+        globalFieldSchema,
+    };
+}
+
+function compareNamed(a: Named<string>, b: Named<string>) {
+    if (a.name < b.name) {
```

Comment:
nit: lift out into a `compareStrings` function in a util file. I'm sure we'll want this again.

Reply:
it's not actually common to need a compareStrings, and if you do need it, its a bit subtle.

If this is the ordering you want (non local aware), the default comparison function for apis like Array.sort will work fine, and if you just need an expression, you can usually just use " < ".

If making a public string comparison, you would want to document how it handles local, Unicode normalization etc, have tests etc. For here all we need is some arbitrary ordering by name.

Also the comparison approach done here is worst case whole string compares: thats not optimal. JS forces some tadoffs there. I'd rather not pick a particular one and claim its the right one. (Ex: other cases might have equal as a common case, and you want to write those comparisons differently so equal only has to do one compare instead of 2)

Reply:
Okay, that makes sense. If the default sort is sufficient, then you could get rid of this function entirely and do a `map(({ name } ) => name).sort()` in each of the places where you're using it (one of them is doing a `map` already). Although, I know that creates an extra array so if we're concerned about performance then maybe that's not ideal.